File: zone-server/src/main/kotlin/net/bestia/zone/economy/shop/OpenShopCMSG.kt
package net.bestia.zone.economy.shop

import net.bestia.bnet.proto.OpenShopCmsgProto
import net.bestia.zone.message.CMSG
import net.bestia.zone.util.EntityId

/**
 * A request for what one merchant has, at the prices of wherever the sender is standing.
 *
 * The merchant chooses whose counter, never whose prices: those come from the sender's own position,
 * because a client that could read a distant town's prices turns the merchant profession into a
 * spreadsheet.
 */
data class OpenShopCMSG(
  override val playerId: Long,
  val merchantEntityId: EntityId,
) : CMSG {

  companion object {
    fun fromBnet(playerId: Long, bnet: OpenShopCmsgProto.OpenShopCMSG): OpenShopCMSG {
      return OpenShopCMSG(playerId, bnet.merchantEntityId)
    }
  }
}

File: bestia-client/src/Bnet/Message/Shop/OpenShopCMSG.cs
using Bnet;

namespace BestiaBehemothClient.Bnet.Message.Shop
{
  /// <summary>
  /// Asks what one merchant has on their shelves. Answered by a ShopOfferSMSG, or by an OperationError
  /// when there is no town underfoot and when the person asked keeps no shop.
  /// </summary>
  /// <remarks>
  /// The merchant chooses whose counter, never whose prices - those come from where we are standing, and
  /// the server re-checks that the merchant is close enough to speak to.
  /// </remarks>
  public partial class OpenShopCMSG : ICMSG
  {
    public ulong MerchantEntityId { get; set; }

    public override Envelope ToEnvelope()
    {
      return new Envelope
      {
        OpenShop = new global::Bnet.OpenShopCMSG
        {
          MerchantEntityId = MerchantEntityId
        }
      };
    }
  }
}
